src/main/components/canvas/mouse-eventlistener.tsx

```diff
@@ -0,0 +1,273 @@
+      layer.addEventListener('mouseup', this.mouseUp);
+    }
+  }
+
+  render() {
+    return (
+      this.state.selectionStarted &&
+      this.state.selectionRectangle.endX && (
+        <svg
+          opacity={0.5}
+          pointerEvents={'none'}
+          style={{
+            position: 'fixed',
+            left: `${Math.min(this.state.selectionRectangle.startX ?? 0, this.state.selectionRectangle.endX ?? 0)}px`,
+            width: `${Math.abs(
+              (this.state.selectionRectangle.startX ?? 0) - (this.state.selectionRectangle.endX ?? 0),
+            )}px`,
+            top: `${Math.min(this.state.selectionRectangle.startY ?? 0, this.state.selectionRectangle.endY ?? 0)}px`,
+            height: `${Math.abs(
+              (this.state.selectionRectangle.startY ?? 0) - (this.state.selectionRectangle.endY ?? 0),
+            )}px`,
+            backgroundColor: '#1E90FF',
+            borderStyle: 'solid',
+            borderWidth: '1px',
+            borderColor: 'blue',
+          }}
+        />
+      )
+    );
+  }
+
+  /**
+   * Mouse down handler for starting the box selection
+   * @param event The triggering mouse down event
+   */
+  private mouseDown = (event: MouseEvent): void => {
+    // if the cursor went out of the bounds of the canvas, then the selection box is still active
+    // we want to continue with the selection box from where we left off
+    if (this.state.selectionStarted) {
+      this.setState((prevState) => {
+        return {
+          ...prevState,
+          selectionRectangle: {
+            ...prevState.selectionRectangle,
+            endX: event.clientX,
+            endY: event.clientY,
+          },
+        };
+      });
+
+      return;
+    }
+
+    // the selection box will activate when clicking anywhere outside the bounds of an element
+    // however:
+    //      * resizing an element can start when clicking slightly outside its bounds
+    //      * the connection/reconnection port of an element is outside its bounding box
+    // in these cases the selection box needs to be disabled
+    if (
+      this.props.resizingInProgress ||
+      this.props.connectingInProgress ||
+      this.props.reconnectingInProgress ||
+      this.props.hoveringInProgress
+    ) {
+      return;
+    }
+
+    this.props.changeSelectionBox(true);
+
+    this.setState({
+      selectionStarted: true,
+      selectionRectangle: {
+        startX: event.clientX,
+        startY: event.clientY,
+        endX: undefined,
+        endY: undefined,
+      },
+    });
+  };
+
+  /**
+   * Mouse up handler for finalising the box selection and determining which elements to select
+   */
+  private mouseUp = (): void => {
+    // if no selection has been started, we can skip determining which
+    // elements are contained in the selection box.
+    if (!this.state.selectionStarted) {
+      return;
+    }
+
+    const selection = this.getElementIDsInSelectionBox();
+    this.props.select(selection);
```

Comment:
why is the selection happening here? shouldn't it happen during mouse move? right now the user doesn't have a feedback of what they are successfully selecting before releasing the cursor, while that way they could immediately see if they are selecting an intended element or not, or whether they are mistakenly selecting unwanted elements. it would of course require calculating which elements should also be deselected.